FlowCrypt/src/main/java/com/flowcrypt/email/jetpack/viewmodel/EkmLoginViewModel.kt

```diff
@@ -0,0 +1,160 @@
+import com.flowcrypt.email.api.retrofit.request.api.DomainRulesRequest
+import com.flowcrypt.email.api.retrofit.request.api.LoginRequest
+import com.flowcrypt.email.api.retrofit.request.model.LoginModel
+import com.flowcrypt.email.api.retrofit.response.api.LoginResponse
+import com.flowcrypt.email.api.retrofit.response.base.ApiError
+import com.flowcrypt.email.api.retrofit.response.base.ApiResponse
+import com.flowcrypt.email.api.retrofit.response.base.Result
+import com.flowcrypt.email.api.retrofit.response.model.OrgRules
+import com.flowcrypt.email.util.exception.EkmNotSupportedException
+import com.flowcrypt.email.util.exception.OrgRulesCombinationNotSupportedException
+import kotlinx.coroutines.launch
+
+/**
+ * This [androidx.lifecycle.ViewModel] can be used to resolve domain rules for enterprise users.
+ *
+ * @author Denis Bondarenko
+ *         Date: 10/23/19
+ *         Time: 12:36 PM
+ *         E-mail: DenBond7@gmail.com
+ */
+class EkmLoginViewModel(application: Application) : BaseAndroidViewModel(application) {
+  private val repository: ApiRepository = FlowcryptApiRepository()
+  val ekmLiveData: MutableLiveData<Result<ApiResponse>?> = MutableLiveData()
+
+  fun fetchPrvKeys(account: String, uuid: String, tokenId: String) {
+    ekmLiveData.value = Result.loading()
+    val context: Context = getApplication()
+
+    viewModelScope.launch {
+      ekmLiveData.value = Result.loading(progressMsg = context.getString(R.string.loading))
+      try {
+        val loginResult = repository.login(
+          context,
+          LoginRequest(ApiName.POST_LOGIN, LoginModel(account, uuid), tokenId)
+        )
+
+        when (loginResult.status) {
+          Result.Status.ERROR, Result.Status.EXCEPTION -> {
+            ekmLiveData.value = loginResult
+            return@launch
+          }
+
+          Result.Status.SUCCESS -> {
+            if (loginResult.data?.isVerified == true) {
+              ekmLiveData.value = Result.loading(
+                progressMsg = context.getString(R.string.loading_domain_rules)
+              )
+              val domainRulesResult = repository.getDomainRules(
+                context,
+                DomainRulesRequest(ApiName.POST_GET_DOMAIN_RULES, LoginModel(account, uuid))
+              )
+
+              if (domainRulesResult.status == Result.Status.SUCCESS) {
+                if (domainRulesResult.data?.orgRules == null) {
+                  ekmLiveData.value = Result.exception(
+                    IllegalArgumentException("${OrgRules::class.java.simpleName} is not specified")
+                  )
+                  return@launch
+                }
+
+                if (!domainRulesResult.data.orgRules.usesKeyManager() ||
+                  !domainRulesResult.data.orgRules.mustAutoImportOrAutoGenPrvWithKeyManager()
+                ) {
+                  ekmLiveData.value = Result.exception(
+                    EkmNotSupportedException(domainRulesResult.data.orgRules)
```

Comment:
I think this is some sort of InvalidStateException? If at this point, you already expected EKM to be in the OrgRules, but here you do a sanity check and it's not there.